File: rust/apps/ensu/src-tauri/src/commands/crypto.rs
use serde::{Deserialize, Serialize};
use tauri::ipc::{InvokeBody, Request, Response};

use crate::commands::common::ApiError;

const CHAT_KEY_HEADER: &str = "chat-key";
const SESSION_UUID_HEADER: &str = "session-uuid";

impl From<ente_ensu_crypto::Error> for ApiError {
    fn from(error: ente_ensu_crypto::Error) -> Self {
        ApiError::new(error.code(), error.to_string())
    }
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EncryptedChatPayload {
    encrypted_data: String,
    header: String,
}

impl From<ente_ensu_crypto::EncryptedChatPayload> for EncryptedChatPayload {
    fn from(payload: ente_ensu_crypto::EncryptedChatPayload) -> Self {
        Self {
            encrypted_data: payload.encrypted_data,
            header: payload.header,
        }
    }
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PayloadEncryptInput {
    value: String,
    key_b64: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PayloadDecryptInput {
    encrypted_data_b64: String,
    header_b64: String,
    key_b64: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldInput {
    value: String,
    key_b64: String,
}

#[tauri::command]
pub fn chat_crypto_generate_key() -> String {
    ente_ensu_crypto::generate_chat_key()
}

#[tauri::command]
pub fn chat_crypto_encrypt_payload(
    input: PayloadEncryptInput,
) -> Result<EncryptedChatPayload, ApiError> {
    ente_ensu_crypto::encrypt_payload(&input.value, &input.key_b64)
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub fn chat_crypto_decrypt_payload(input: PayloadDecryptInput) -> Result<String, ApiError> {
    ente_ensu_crypto::decrypt_payload(&input.encrypted_data_b64, &input.header_b64, &input.key_b64)
        .map_err(Into::into)
}

#[tauri::command]
pub fn chat_crypto_encrypt_field(input: FieldInput) -> Result<String, ApiError> {
    ente_ensu_crypto::encrypt_field_b64(&input.value, &input.key_b64).map_err(Into::into)
}

#[tauri::command]
pub fn chat_crypto_decrypt_field(input: FieldInput) -> Result<String, ApiError> {
    ente_ensu_crypto::decrypt_field_b64(&input.value, &input.key_b64).map_err(Into::into)
}

#[tauri::command]
pub fn chat_crypto_encrypt_attachment(request: Request<'_>) -> Result<Response, ApiError> {
    transform_attachment(request, ente_ensu_crypto::encrypt_attachment)
}

#[tauri::command]
pub fn chat_crypto_decrypt_attachment(request: Request<'_>) -> Result<Response, ApiError> {
    transform_attachment(request, ente_ensu_crypto::decrypt_attachment)
}

fn transform_attachment(
    request: Request<'_>,
    transform: impl FnOnce(&[u8], &str, &str) -> ente_ensu_crypto::Result<Vec<u8>>,
) -> Result<Response, ApiError> {
    let InvokeBody::Raw(data) = request.body() else {
        return Err(ApiError::new("invalid_args", "Expected attachment bytes"));
    };
    let header = |name| {
        request
            .headers()
            .get(name)
            .and_then(|value| value.to_str().ok())
            .ok_or_else(|| ApiError::new("invalid_args", format!("Missing {name} header")))
    };
    transform(data, header(CHAT_KEY_HEADER)?, header(SESSION_UUID_HEADER)?)
        .map(Response::new)
        .map_err(Into::into)
}

File: web/apps/ensu/src/services/chat/crypto.ts
import { isTauriRuntime } from "@/services/tauri-runtime";
import * as wasmCrypto from "ente-ensu-wasm";

type TauriCoreModule = typeof import("@tauri-apps/api/core");

export interface EncryptedChatPayload {
    encryptedData: string;
    header: string;
}

const tauriInvoke = async <T>(
    ...args: Parameters<TauriCoreModule["invoke"]>
) => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<T>(...args);
};

const invokeAttachmentCommand = async (
    command: string,
    data: Uint8Array,
    chatKeyB64: string,
    sessionUuid: string,
) =>
    new Uint8Array(
        await tauriInvoke<ArrayBuffer>(command, data, {
            headers: { "chat-key": chatKeyB64, "session-uuid": sessionUuid },
        }),
    );

export const generateChatKey = async () =>
    isTauriRuntime()
        ? tauriInvoke<string>("chat_crypto_generate_key")
        : wasmCrypto.generateChatKey();

export const encryptChatPayload = async (
    payload: unknown,
    chatKeyB64: string,
): Promise<EncryptedChatPayload> => {
    const value = JSON.stringify(payload);
    if (isTauriRuntime()) {
        return tauriInvoke("chat_crypto_encrypt_payload", {
            input: { value, keyB64: chatKeyB64 },
        });
    }
    return wasmCrypto.encryptChatPayload(value, chatKeyB64);
};

export const decryptChatPayload = async (
    { encryptedData, header }: EncryptedChatPayload,
    chatKeyB64: string,
): Promise<unknown> => {
    const value = isTauriRuntime()
        ? await tauriInvoke<string>("chat_crypto_decrypt_payload", {
              input: {
                  encryptedDataB64: encryptedData,
                  headerB64: header,
                  keyB64: chatKeyB64,
              },
          })
        : await wasmCrypto.decryptChatPayload(
              encryptedData,
              header,
              chatKeyB64,
          );
    return JSON.parse(value);
};

export const encryptChatField = async (
    value: string,
    chatKeyB64: string,
): Promise<string> =>
    isTauriRuntime()
        ? tauriInvoke("chat_crypto_encrypt_field", {
              input: { value, keyB64: chatKeyB64 },
          })
        : wasmCrypto.encryptChatField(value, chatKeyB64);

export const decryptChatField = async (
    value: string,
    chatKeyB64: string,
): Promise<string> =>
    isTauriRuntime()
        ? tauriInvoke("chat_crypto_decrypt_field", {
              input: { value, keyB64: chatKeyB64 },
          })
        : wasmCrypto.decryptChatField(value, chatKeyB64);

export const encryptAttachmentBytes = async (
    data: Uint8Array,
    chatKeyB64: string,
    sessionUuid: string,
): Promise<Uint8Array> =>
    isTauriRuntime()
        ? invokeAttachmentCommand(
              "chat_crypto_encrypt_attachment",
              data,
              chatKeyB64,
              sessionUuid,
          )
        : wasmCrypto.encryptChatAttachment(data, chatKeyB64, sessionUuid);

export const decryptAttachmentBytes = async (
    data: Uint8Array,
    chatKeyB64: string,
    sessionUuid: string,
): Promise<Uint8Array> =>
    isTauriRuntime()
        ? invokeAttachmentCommand(
              "chat_crypto_decrypt_attachment",
              data,
              chatKeyB64,
              sessionUuid,
          )
        : wasmCrypto.decryptChatAttachment(data, chatKeyB64, sessionUuid);
